### ISSUE TITLE
The `punycode` module is deprecated in Node.js 21 (type = module)

### ISSUE THREAD

**Issue by adriaandotcom** [openai/openai-node#527]
### Confirm this is a Node library issue and not an underlying OpenAI API issue

- [X] This is an issue with the Node library

### Describe the bug

When importing the `openai` package in Node.js 21 it throws an error:

```
(node:38591) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
    at node:punycode:3:9
    at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/realm:392:7)
    at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/realm:328:10)
    at loadBuiltinModule (node:internal/modules/helpers:101:7)
    at Module._load (node:internal/modules/cjs/loader:1001:17)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/Users/adriaan/app/node_modules/whatwg-url/lib/url-state-machine.js:2:18)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
```

### To Reproduce

`package.json`:

```json
{
  "main": "index.js",
  "type": "module",
  "engines": {
    "node": "21.2.0"
  },
  "scripts": {
    "dev": "node --trace-deprecation -r dotenv/config index.js dotenv_config_path=.env.dev"
  },
  "dependencies": {
    "openai": "^4.19.1"
  },
  "devDependencies": {
    "dotenv": "^16.3.1"
  }
}
```

`index.js`:

```js
import OpenAI from "openai";

console.log("hi");
```

`.env.dev`:

```
OPENAI_API_KEY=
ASSISTANT_ID=
PORT=3011
BASE_URL=http://localhost:3011
```

### OS

macOS

### Node version

v21.2.0

### Library version

openai 4.19.1

**Comment by rattrayalex** [comment]
Sorry about that; thanks for reporting. 

This looks to be coming from `whatwg-url@5.x`, which is a dependency of `node-fetch@2.x`. This [has been fixed in newer versions of whatwg-url](https://github.com/jsdom/whatwg-url/issues/261), but `node-fetch` v2.x would need to update its deps, and I don't think that's likely (node-fetch has been at v3 for some time, but this library can't use it because it supports CommonJS). 

The good news is we're planning to move from `node-fetch` to `undici`, so this would be fixed by https://github.com/openai/openai-node/pull/402. 

That'll probably be a breaking change, so it won't come out til we're able to do a major version bump.

In the short term, you can work around this by overriding the dependency in your `package.json`, eg: 

```json
"overrides": {
  "whatwg-url": "13.0.0"
}
```

**Comment by adriaanvanrossum** [comment]
Thanks for the help! I moved back to Node.js 20 for now.

What also could be a solution is to use native [Node.js fetch from version 21](https://nodejs.org/en/blog/announcements/v21-release-announce). It's not experimental anymore and included in the default distributions.

**Comment by philkunz** [comment]
This is still an issue. Caused a bit of an explanation at a client who saw the logs that openai itself is using deprecated imports in the dependency chain.

**Comment by rattrayalex** [comment]
Sorry about that. This issue will be fixed by https://github.com/openai/openai-node/pull/402 and won't be tackled before, so I'll close this as a duplicate.

**Comment by dgellow** [comment]
An option would be to rely on a patch utility such as https://github.com/ds300/patch-package to patch `whatwg-url`. If we replace the `require("punycode")` by `require("punycode/")` and add punycode as a dependency that should get rid of the warning.

**Comment by philkunz** [comment]
@dgellow Please don't! Modules are not supposed to be patched like this. It will create all sorts of problems, especially with pnpm, where modules are linked instead of installed per project. The right choice is not to import a fetch polyfill, when there is fetch natively, and choose a good polyfill like @adobe/fetch instead of node-fetch only if needed when running in an outdated environment. Node-fetch by now is the new request module. Just bloated and has problems with memory leaks.

Last not but least: It is kinda funny, that a company like openai, with something as smart as GPT 4 can't produce a proper isomorphic TypeScript client, but instead relies on node.js specific APIs.

**Comment by dgellow** [comment]
Actually pnpm supports that patching feature natively via `pnpm patch`: https://pnpm.io/cli/patch.

**Comment by rattrayalex** [comment]
As [noted above](https://github.com/openai/openai-node/issues/527#issuecomment-1823471697), we are working on moving to native `fetch`/`undici`, which was still experimental in some supported Node runtimes when this library was released (hence using `node-fetch`).

We also document a non-`patch`-based workaround just above as well.

**Comment by DanteAndroid** [comment]
> Sorry about that; thanks for reporting.
> 
> This looks to be coming from `whatwg-url@5.x`, which is a dependency of `node-fetch@2.x`. This [has been fixed in newer versions of whatwg-url](https://github.com/jsdom/whatwg-url/issues/261), but `node-fetch` v2.x would need to update its deps, and I don't think that's likely (node-fetch has been at v3 for some time, but this library can't use it because it supports CommonJS).
> 
> The good news is we're planning to move from `node-fetch` to `undici`, so this would be fixed by #402.
> 
> That'll probably be a breaking change, so it won't come out til we're able to do a major version bump.
> 
> In the short term, you can work around this by overriding the dependency in your `package.json`, eg:
> 
> ```json
> "overrides": {
>   "whatwg-url": "13.0.0"
> }
> ```
<img width="449" alt="image" src="https://github.com/user-attachments/assets/97c004c9-ca6f-4ed3-bd9c-28ef5f3fd53d">

Hey, this does not work for me, why?
Here is my package.json in root directory.
Thanks!
```
{
    "overrides": {
        "whatwg-url": "13.0.0"
      }
}
```

**Comment by lonbinder** [comment]
> > Sorry about that; thanks for reporting.
> > This looks to be coming from `whatwg-url@5.x`, which is a dependency of `node-fetch@2.x`. This [has been fixed in newer versions of whatwg-url](https://github.com/jsdom/whatwg-url/issues/261), but `node-fetch` v2.x would need to update its deps, and I don't think that's likely (node-fetch has been at v3 for some time, but this library can't use it because it supports CommonJS).
> > The good news is we're planning to move from `node-fetch` to `undici`, so this would be fixed by #402.
> > That'll probably be a breaking change, so it won't come out til we're able to do a major version bump.
> > In the short term, you can work around this by overriding the dependency in your `package.json`, eg:
> > ```json
> > "overrides": {
> >   "whatwg-url": "13.0.0"
> > }
> > ```
> 
> <img alt="image" width="449" src="https://private-user-images.githubusercontent.com/10005563/357316571-97c004c9-ca6f-4ed3-bd9c-28ef5f3fd53d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4MzUyMzYsIm5iZiI6MTcyODgzNDkzNiwicGF0aCI6Ii8xMDAwNTU2My8zNTczMTY1NzEtOTdjMDA0YzktY2E2Zi00ZWQzLWJkOWMtMjhlZjVmM2ZkNTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDE1NTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3NTFhY2NiMDMzNGIyNzhmYjlmYmRmZDA3YjE2MzA1MjAyMDFjZmU2NTM0YmQwOWIyOGNjOTQ0MjMzMDQwYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FZEknDI8mPchq6fS_4LL4YqG0sftg7iFRtUJlHaIkYk">
> Hey, this does not work for me, why? Here is my package.json in root directory. Thanks!
> 
> ```
> {
>     "overrides": {
>         "whatwg-url": "13.0.0"
>       }
> }
> ```

Make sure you run `npm update whatwg-url` after changing your package.json

**Comment by oGabrielFreitas** [comment]
> > > Sorry about that; thanks for reporting.
> > > This looks to be coming from `whatwg-url@5.x`, which is a dependency of `node-fetch@2.x`. This [has been fixed in newer versions of whatwg-url](https://github.com/jsdom/whatwg-url/issues/261), but `node-fetch` v2.x would need to update its deps, and I don't think that's likely (node-fetch has been at v3 for some time, but this library can't use it because it supports CommonJS).
> > > The good news is we're planning to move from `node-fetch` to `undici`, so this would be fixed by #402.
> > > That'll probably be a breaking change, so it won't come out til we're able to do a major version bump.
> > > In the short term, you can work around this by overriding the dependency in your `package.json`, eg:
> > > ```json
> > > "overrides": {
> > >   "whatwg-url": "13.0.0"
> > > }
> > > ```
> > 
> > 
> > <img alt="image" width="449" src="https://private-user-images.githubusercontent.com/10005563/357316571-97c004c9-ca6f-4ed3-bd9c-28ef5f3fd53d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4MzUyMzYsIm5iZiI6MTcyODgzNDkzNiwicGF0aCI6Ii8xMDAwNTU2My8zNTczMTY1NzEtOTdjMDA0YzktY2E2Zi00ZWQzLWJkOWMtMjhlZjVmM2ZkNTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDE1NTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3NTFhY2NiMDMzNGIyNzhmYjlmYmRmZDA3YjE2MzA1MjAyMDFjZmU2NTM0YmQwOWIyOGNjOTQ0MjMzMDQwYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FZEknDI8mPchq6fS_4LL4YqG0sftg7iFRtUJlHaIkYk">
> > Hey, this does not work for me, why? Here is my package.json in root directory. Thanks!
> > ```
> > {
> >     "overrides": {
> >         "whatwg-url": "13.0.0"
> >       }
> > }
> > ```
> 
> Make sure you run `npm update whatwg-url` after changing your package.json

Worked, thanks.

**Comment by DavidFarago** [comment]
Overriding 2whatwg-url" worked for me too (the last two days), but today I get a punycode deprecation warning again:
```
(node:3365860) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
```

**Comment by sagarkpatil3** [comment]
> > > Sorry about that; thanks for reporting.
> > > This looks to be coming from `whatwg-url@5.x`, which is a dependency of `node-fetch@2.x`. This [has been fixed in newer versions of whatwg-url](https://github.com/jsdom/whatwg-url/issues/261), but `node-fetch` v2.x would need to update its deps, and I don't think that's likely (node-fetch has been at v3 for some time, but this library can't use it because it supports CommonJS).
> > > The good news is we're planning to move from `node-fetch` to `undici`, so this would be fixed by #402.
> > > That'll probably be a breaking change, so it won't come out til we're able to do a major version bump.
> > > In the short term, you can work around this by overriding the dependency in your `package.json`, eg:
> > > ```json
> > > "overrides": {
> > >   "whatwg-url": "13.0.0"
> > > }
> > > ```
> > 
> > 
> > <img alt="image" width="449" src="https://private-user-images.githubusercontent.com/10005563/357316571-97c004c9-ca6f-4ed3-bd9c-28ef5f3fd53d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4MzUyMzYsIm5iZiI6MTcyODgzNDkzNiwicGF0aCI6Ii8xMDAwNTU2My8zNTczMTY1NzEtOTdjMDA0YzktY2E2Zi00ZWQzLWJkOWMtMjhlZjVmM2ZkNTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDE1NTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3NTFhY2NiMDMzNGIyNzhmYjlmYmRmZDA3YjE2MzA1MjAyMDFjZmU2NTM0YmQwOWIyOGNjOTQ0MjMzMDQwYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FZEknDI8mPchq6fS_4LL4YqG0sftg7iFRtUJlHaIkYk">
> > Hey, this does not work for me, why? Here is my package.json in root directory. Thanks!
> > ```
> > {
> >     "overrides": {
> >         "whatwg-url": "13.0.0"
> >       }
> > }
> > ```
> 
> Make sure you run `npm update whatwg-url` after changing your package.json

you saved my life
